### ISSUE TITLE
Emit logs from <match **> not working

### ISSUE THREAD

**Issue by g3kr** [fluent/fluent-plugin-rewrite-tag-filter#94]
Mine is a simple setup

```
<source>
  @type dummy
  tag dummy
  dummy [
     {
       "container_id": "51b7eb1860ab901975cb3b43fc2116c597a9011c1ff7ff38b403e3005338e562",
       "container_name": "/ecs-Capillaryreadingtest-capillary-48-Capillaryreadingtest-capillary-b0e496d6ea96d4cf3800",
       "contextMap": {},
       "level": "INFO",
       "loggerFqcn": "org.apache.log4j.Category",
       "loggerName": "com.qd.bi.test.api.v1.ZooKeeperV1",
       "message": "Received.",
       "requestId": "8uhe5308-6203-4085-9077-8486637048b3",
       "source": "stdout",
       "thread": "http-nio-8080-exec-7",
       "threadId": 7,
       "threadPriority": 5,
       "foo":"bar"
   }
  ]
</source>

<match **>
  @type rewrite_tag_filter
  <rule>
    key $.foo
    pattern ^(.+)$
    tag normalMsg
  </rule>
</match>

<match normalMsg>
  @type stdout
</match>
```

What I observe though is rewrite_tag_filter is not able to emit logs with the `<match **>` construct. However it works with `<match dummy>` construct if i include a `tag dummy` in the `source`. My source does not contain one specific tag and its difficult for me to include a tag in that case. How do I get this to work with `<match **>`

**Comment by repeatedly** [comment]
This case https://docs.fluentd.org/configuration/config-file#note-on-match-order ?

**Comment by AlexOQ** [comment]
I don't know if it's related somehow, but I too have a problem with log events with the changed tag.
```
<match kubernetes.var.log.containers.**>
  @type rewrite_tag_filter
  <rule>
    key some_key
    pattern ^some_value$
    tag wanted.stuff
  </rule>
  <rule>
    key some_key
    pattern ^some_value$
    invert true
    tag unmatched.${tag}
  </rule>
</match>

<match wanted.stuff>
  @type elasticsearch                  # doesn't work with stdout or file as well
  index wanted.stuff.new.index
  include es.conf
</match>

<match **>
  @type elasticsearch
  index existing.index
  include es.conf
</match>
```

What happens is that `unmatched.**` get rerouted and on to Kibana successfully, but `wanted.stuff` never appears. Tried with file and stdout as well. Elastic config is the same, save for index name, which in my case is a new index (have not created manually). Can confirm Elastic creates new indices every day for the existing,and some other configs as well, `auto_create_index` is persistent on. Here the only difference is the fact that I use rewrite_tag_filter.
Any help is appreciated.

**Comment by kvokka** [comment]
Facing the same issue

@AlexOQ Can you please share how you solved your issue?

**Comment by AlexOQ** [comment]
Since it was a past work project I don't have access to the exact configuration, but it was along the following logic.
You use `<match>` to filter for `wanted` and `rest`. Also works with multiple `wanted` rules. If you use `<filter>` you'll lose filtered out logs, that's why `@LABEL` is the way. Don't forget to tag with prefixes for easier catching/filtering later down the line and in Kibana index management.
Apply `@LABEL` for each of `wanted` filters, including the `everything_else`. Then in the same config apply `@LABEL` handling logic.
The problem I've found using only `<match>` is that when you overwrite a tag of a log, you cannot catch it again using a new prefix because I was not rethrowing the log back to the pipeline to try and filter it as a new prefixed log.

Basically, rewrite_tag_filter works as it should, I was just expecting double-handling during a single log pass. The case works with `@LABEL` because labels are handled after matches.

**Comment by kvokka** [comment]
Thank you @AlexOQ for such a quick and full response. Will stick with the match & labels approach then and keep tags alone.

**Comment by AlexOQ** [comment]
Rewriting tags provided for easier index management in Kibana, IMO. But that is ultimately a personal preference.
I should close the issue anyway because the filter plugin works as intended.

**Comment by Watson1978** [comment]
Thank you for the report, and sorry for the very long delay.

This is not a bug in the plugin. It is Fluentd's match order combined with this plugin's loop prevention.

In the configuration in the first post, `<match **>` is the first `<match>` in the file, so it catches every event, including the ones this plugin re-emits:

1. `dummy` is emitted by the input and caught by `<match **>`.
2. The rule matches, so the event is re-emitted with the tag `normalMsg`.
3. A re-emitted event is routed from the top of the routing table again, and `<match **>` still matches `normalMsg`, so it comes back into `rewrite_tag_filter`.
4. The same rule matches again and the resulting tag is `normalMsg`, which is identical to the incoming tag. Emitting it again would repeat step 3 forever, so the plugin skips the event. Only a trace-level log is written, which is why it looks like the events silently disappear. You can see it with `@log_level trace` on that `<match>`.

So `<match normalMsg>` is never reached. It works with `<match dummy>` because `normalMsg` no longer matches the first `<match>` and falls through to `<match normalMsg>`. This is exactly the behavior @repeatedly linked to: https://docs.fluentd.org/configuration/config-file#note-on-match-order

Note that every Fluentd event always has a tag, so you do not need to give the input a single fixed tag. You only need the outputs for the rewritten tags to be reachable. There are two ways to do that.

Option 1: put the specific `<match>` before the catch-all, since Fluentd uses the first matching `<match>`:

````
<source>
  @type sample
  tag dummy
  sample {"foo":"bar"}
</source>

<match normalMsg>
  @type stdout
</match>

<match **>
  @type rewrite_tag_filter
  <rule>
    key $.foo
    pattern /^(.+)$/
    tag normalMsg
  </rule>
</match>
````

Option 2: route the re-emitted events to a label with the per-rule `label` parameter (available since v2.4.0). Events sent to a label do not go back through the top-level `<match>` chain, so the order no longer matters:

````
<source>
  @type sample
  tag dummy
  sample {"foo":"bar"}
</source>

<match **>
  @type rewrite_tag_filter
  <rule>
    key $.foo
    pattern /^(.+)$/
    tag normalMsg
    label @NORMAL
  </rule>
</match>

<label @NORMAL>
  <match normalMsg>
    @type stdout
  </match>
</label>
````

Regarding the follow-up case with `kubernetes.var.log.containers.**`: `wanted.stuff` is not covered by that pattern, so those events should have reached `<match wanted.stuff>`. The fact that everything ended up in `unmatched.**` instead suggests the first rule never fired. Two things to check. First, if the field is nested, `key some_key` does not reach it and you need the record accessor syntax, for example `key $.kubernetes.labels.some_key`. Second, a field that is missing from the record is read as an empty string, and an inverted rule fires in that case because an empty string does not match `^some_value$`, so a missing or misspelled key sends every record to the inverted rule. Both cases produce exactly the symptom described there.

Since this is expected behavior rather than a defect, and the discussion already converged on the label-based approach, I am closing this issue. The relevant behavior is now described in the README, in the sections about unmatched records and about altering labels. Please feel free to open a new issue if you hit something that these two options do not cover.
